src/messaging/ExchangeContext.cpp

```diff
@@ -336,7 +364,6 @@ ExchangeContext::ExchangeContext(ExchangeManager * em, uint16_t ExchangeId, cons
 ExchangeContext::~ExchangeContext()
 {
     VerifyOrDie(mExchangeMgr != nullptr && GetReferenceCount() == 0);
-    VerifyOrDie(!IsAckPending());
```

Comment:
I had to remove this to get the test to pass. I'm not sure why this is here @bzbarsky-apple ?

Reply:
If this assert is failing then ***someone did not close the exchange properly and it leaked***.  Probably because of the rules change for failing SendMessage under OnMessageReceived.

We have no other checks for exchanges leaking, if I recall correctly; the exchange manager does not assert that none leaked.  So if this is failing after this PR, then this PR introduced a leak.

Reply:
How exactly does it leak? Can you shed more details?

Reply:
> How exactly does it leak? Can you shed more details?

I'd have to debug to tell exactly what is leaking and how, but the upshot is that we are destroying an exchange (probably by destroying the pool it was allocated from) without having actually closed it (and therefore destroyed it via the normal "refcount goes to 0" codepath).

Reply:
So, the sequence of events in the unit-test that necessitated this was as follows:

1. Responder receives a message on an exchange.
2. Responder attempts to send a message on the exchange. We artificially inject an error that results in `SendMessage` failing.
3. That eventually results in the `ExchangeHolder` aborting the EC (since a failure to send means it is responsible to abort the EC). That results in a call to `DoClose`, which calls `FlushAcks`. That _also_ fails since we still have the 'inject failure' enabled on the EC. That prevents us from getting to `ExchangeMessageDispatch::SendMessage`, which is what would have cleared that ACK bit by calling `TakePendingPeerAckMessageCounter`.
4. The EC's destructor is called eventually, and VerifyOrDie triggers.

Reply:
Causing a session to be released before calling `SendMessage` has the same resultant effect btw.

Reply:
> Causing a session to be released before calling SendMessage has the same resultant effect btw.

Hmm.  Maybe we should clear the "ack pending" state if we try to send an ack and fail....  Because at that point we won't be able to send it sanely.

Reply:
Since the crux of what this was checking was that `DoClose` got called before we got here, checking for the presence of the Closed flag is a more direct, clearer, and more encompassing check.

@bzbarsky-apple was good with this suggestion, so closing comment.